mixer/pkg/log/log.go

```diff
@@ -0,0 +1,244 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package log provides the canonical logging functionality used by go-based
+// Istio components.
+//
+// Istio's logging subsystem is built on top of the [Zap](https://godoc.org/go.uber.org/zap) package.
+// High performance scenarios should use the Error, Warn, Info, and Debug methods. Lower perf
+// scenarios can use the more expensive convenience methods such as Debugf and Warnw.
+//
+// The package provides direct integration with the Cobra command-line processor so make it
+// easy to build programs that use a consistent interface for logging. Here's an example
+// of a simple Cobra-based program using this log package:
+//
+//		func main() {
+//			// get the default logging options
+//			options := log.NewOptions()
+//
+//			rootCmd := &cobra.Command{
+//				Run: func(cmd *cobra.Command, args []string) {
+//
+//					// configure the logging system
+//					log.Configure(options)
+//
+//					// output some logs
+//					log.Info("Hello")
+//					log.Sync()
+//				},
+//			}
+//
+//			// add logging-specific flags to the cobra command
+//			options.AttachCobraFlags(rootCmd)
+//			rootCmd.SetArgs(os.Args[1:])
+//			rootCmd.Execute()
+//		}
+//
+// Once configured, this package intercepts the output of the standard golang "log" package as well as anything
+// sent to the global zap logger (zap.L()).
+package log
+
+import (
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+)
+
+// global variables against which all our logging occurs
+var logger *zap.Logger = zap.NewNop()
+var sugar *zap.SugaredLogger = logger.Sugar()
+
+// Configure initializes Istio's logging subsystem.
+//
+// You typically call this once at process startup.
+// Once this call returns, the logging system is ready to accept data.
+func Configure(options *Options) error {
+	zapConfig := zap.Config{
+		Level:       zap.NewAtomicLevelAt(zapcore.DebugLevel), // enable everything since we filter ourselves
+		Development: false,
+
+		Sampling: &zap.SamplingConfig{
+			Initial:    100,
+			Thereafter: 100,
+		},
+
+		Encoding: "console",
+		EncoderConfig: zapcore.EncoderConfig{
+			TimeKey:        "time",
+			LevelKey:       "level",
+			NameKey:        "logger",
+			CallerKey:      "caller",
+			MessageKey:     "msg",
+			StacktraceKey:  "stack",
+			LineEnding:     zapcore.DefaultLineEnding,
+			EncodeLevel:    zapcore.LowercaseLevelEncoder,
+			EncodeCaller:   zapcore.ShortCallerEncoder,
+			EncodeTime:     zapcore.ISO8601TimeEncoder,
+			EncodeDuration: zapcore.StringDurationEncoder,
+		},
+
+		OutputPaths:       options.OutputPaths,
+		ErrorOutputPaths:  []string{"stderr"},
+		DisableCaller:     true,
+		DisableStacktrace: true,
```

Comment:
Is there no way to log the stack trace automagically ?

Reply:
I didn't provide that option. This can easily be plumbed through.